pkg/cloud/user_credentials.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	p := c.cs.Account.NewListAccountsParams()
+	p.SetDomainid(account.Domain.ID)
+	setIfNotEmpty(account.ID, p.SetId)
+	setIfNotEmpty(account.Name, p.SetName)
+	resp, retErr := c.cs.Account.ListAccounts(p)
+	if retErr != nil {
+		return retErr
+	} else if resp.Count != 1 {
+		return errors.Errorf("expected 1 Account with account name %s in domain ID %s, but got %d",
+			account.Name, account.Domain.ID, resp.Count)
+	}
+	account.ID = resp.Accounts[0].Id
+	account.Name = resp.Accounts[0].Name
+	return nil
+}
+
+// ResolveUser resolves a user's information.
+func (c *client) ResolveUser(user *User) error {
+	// Resolve account prior to any user resolution activity.
+	if err := c.ResolveAccount(&user.Account); err != nil {
+		return errors.Wrap(err, "error encountered when resolving account details")
+	}
+
+	p := c.cs.User.NewListUsersParams()
+	p.SetAccount(user.Account.Name)
+	p.SetDomainid(user.Domain.ID)
+	p.SetListall(true)
+	resp, err := c.cs.User.ListUsers(p)
+	if err != nil {
+		return err
+	} else if resp.Count != 1 {
+		return errors.Errorf("expected 1 User with username %s but got %d", user.Name, resp.Count)
+	}
+
+	user.ID = resp.Users[0].Id
+	user.Name = resp.Users[0].Username
+
+	return nil
+}
+
+// ResolveUserKeys resolves a user's api keys.
+func (c *client) ResolveUserKeys(user *User) error {
+	// Resolve user prior to any api key resolution activity.
+	if err := c.ResolveUser(user); err != nil {
+		return errors.Wrap(err, "error encountered when resolving user details")
+	}
+
+	p := c.cs.User.NewGetUserKeysParams(user.ID)
+	resp, err := c.cs.User.GetUserKeys(p)
+	if err != nil {
+		return errors.Errorf("error encountered when resolving user api keys for user %s", user.Name)
+	}
+	user.APIKey = resp.Apikey
+	user.SecretKey = resp.Secretkey
+	return nil
+}
+
+// GetUserWithKeys will search a domain and account for the first user that has api keys.
+// Returns true if a user is found and false otherwise.
+func (c *client) GetUserWithKeys(user *User) (bool, error) {
+	// Resolve account prior to any user resolution activity.
+	if err := c.ResolveAccount(&user.Account); err != nil {
+		return false, errors.Wrap(err, "error encountered when resolving account details")
+	}
+
+	// List users and take first user that has already has api keys.
+	p := c.cs.User.NewListUsersParams()
+	p.SetAccount(user.Account.Name)
+	p.SetDomainid(user.Domain.ID)
+	p.SetListall(true)
+	resp, err := c.cs.User.ListUsers(p)
+	if err != nil {
+		return false, err
+	}
+
+	// Return first user with keys.
+	for _, possibleUser := range resp.Users {
+		user.ID = possibleUser.Id
+		if err := c.ResolveUserKeys(user); err == nil {
+			return true, nil
+		}
+	}
+	user.ID = ""
+	return false, nil
+}
+
+// GetOrCreateUserWithKeys will search a domain and account for the first user that has api keys.
+// Creates a CAPC user if no such user is found.
+func (c *client) GetOrCreateUserWithKeys(user *User) error {
+	return nil
+}
```

Comment:
Well, this is a draft PR, so yes I'll remove them. : )

Reply:
BTW, I do eventually want methods like this in the codebase to be able to create domains/accounts/users on the fly in testing.

But since we will probably never do that in production, I'll probably put them in the test suite itself.